src/polyswarmd/websockets/filter.py

```diff
@@ -0,0 +1,166 @@
+            gevent.sleep(wait)
+            # Spawn the next version of this instance
+            greenlet = gevent.spawn(self.get_new_entries)
+            try:
+                result = greenlet.get(block=True, timeout=self.MAX_WAIT)
+            # KeyError generally arises when the JSONSchema describing a message is fed an invalid value.
+            except KeyError as e:
+                logger.exception(e)
+                continue
+            # ConnectionError generally occurs when we cannot fetch events
+            except (ConnectionError, gevent.Timeout):
+                logger.exception("Error thrown in get_new_entries")
+                wait = self.compute_wait(ctr + 2)
+                continue
+
+            # Reset the ctr if we recieved a non-empty response or we shouldn't backoff
+            if len(result) != 0:
+                callback(result)
+                ctr = 0
+
+            # We add gaussian randomness so that requests are queued all-at-once.
+            wait = self.compute_wait(ctr)
+            logger.debug("%s wait=%f", self.contract_event_name(), wait)
+
+    def __hash__(self):
+        return hash(self.filter)
+
+
+class FilterManager():
+    """Manages access to filtered Ethereum events."""
+
+    wrappers: Collection[FilterWrapper]
+    pool: Group
+
+    def __init__(self):
+        self.wrappers = set()
+        self.pool = Group()
+
+    def register(self, flt: LogFilter, fmt_cls: Type[messages.EventLogMessage] = lambda x: x, backoff=True):
+        "Add a new filter, with an optional associated WebsocketMessage-serializer class"
+        wrapper = FilterWrapper(flt, fmt_cls, backoff)
+        self.wrappers.add(wrapper)
+        logger.debug('Registered new filter: %s', wrapper)
+
+    def flush(self):
+        "End all event polling, uninstall all filters and remove their corresponding wrappers"
+        self.pool.kill()
+        logger.debug('Flushing %d filters', len(self.wrappers))
+        for filt in self.wrappers:
+            filt.uninstall()
+        self.wrappers.clear()
+
+    @contextmanager
+    def fetch(self):
+        "Return a queue of currently managed contract events"
+        try:
+            queue = Queue()
+            # Greenlet's can continue to exist beyond the lifespan of
+            # the object itself. Failing to use a weakref here can prevent filters
+            # destructors from running
+            for wrapper in map(weakref.proxy, self.wrappers):
+                self.pool.spawn(wrapper.spawn_poll_loop, queue.put_nowait)
+
+            yield queue
+        finally:
+            self.flush()
+
+    def setup_event_filters(self, chain):
+        "Setup the most common event filters"
+        if len(self.wrappers) != 0:
+            logger.exception("Attempting to initialize already initialized filter manager")
+            return
+
+        # Setup Latest
+        self.register(chain.w3.eth.filter('latest'), messages.LatestEvent.make(chain.w3.eth))
```

Comment:
Backoff should be false here as well, though in an ideal world these should be consistent